File: 555/555.rb
# Given the root of a binary tree, return its maximum depth.

# A binary tree's maximum depth is the number of nodes along the 
# longest path from the root node down to the farthest leaf node.

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
    return depth_first_search(root, 0)
end

def depth_first_traversal(node, level)
    return level if node == nil
    left = depth_first_traversal(node.left, level + 1)
    right = depth_first_traversal(node.right, level + 1)
    return [left, right].max
end

File: leetcode/555/555.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        '''
        Given the root of a binary tree, return its maximum depth.

        A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.
        '''
        return self.depthFirst(root)
        
    def depthFirst(self, node: TreeNode) -> int:
        if node is None:
            return 0
        left = self.depthFirst(node.left)
        right = self.depthFirst(node.right)
        return 1 + max(left, right)

